lib/rubocop/cop/rspec/verified_double_reference.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for consistent verified double reference style.
+      #
+      # Only investigates references that are one of the supported styles.
+      #
+      # @see https://relishapp.com/rspec/rspec-mocks/docs/verifying-doubles
+      #
+      # This cop can be configured in your configuration using the
+      # `EnforcedStyle` option and supports `--auto-gen-config`.
+      #
+      # @example `EnforcedStyle: constant`
+      #   # bad
+      #   let(:foo) do
+      #     instance_double('ClassName', method_name: 'returned_value')
+      #   end
+      #
+      #   # good
+      #   let(:foo) do
+      #     instance_double(ClassName, method_name: 'returned_value')
+      #   end
+      #
+      # @example `EnforcedStyle: string`
+      #   # bad
+      #   let(:foo) do
+      #     instance_double(ClassName, method_name: 'returned_value')
+      #   end
+      #
+      #   # good
+      #   let(:foo) do
+      #     instance_double('ClassName', method_name: 'returned_value')
+      #   end
+      #
+      # @example Reference is not in the supported style list. No enforcement
+      #
+      #   # good
+      #   let(:foo) do
+      #     instance_double(@klass, method_name: 'returned_value')
+      #   end
+      class VerifiedDoubleReference < Base
+        extend AutoCorrector
+        include ConfigurableEnforcedStyle
+
+        MSG = 'Prefer `%<style>s` references for verified doubles.'
+
+        RESTRICT_ON_SEND = Set[
+          :class_double,
+          :class_spy,
+          :instance_double,
+          :instance_spy,
+          :mock_model,
+          :object_double,
+          :object_spy,
+          :stub_model
+        ].freeze
+
+        REFERENCE_TYPE_STYLES = {
+          str: :string,
+          const: :constant
+        }.freeze
+
+        # @!method verified_double(node)
+        def_node_matcher :verified_double, <<~PATTERN
+          (send
+            nil?
+            RESTRICT_ON_SEND
+            $_class_reference
+            ...)
+        PATTERN
+
+        def on_send(node)
+          verified_double(node) do |class_reference|
+            return correct_style_detected if preferred_style?(class_reference)
+
+            message = format(MSG, style: style)
+            expression = class_reference.loc.expression
+
+            add_offense(expression, message: message) do |corrector|
+              violation = class_reference.children.last.to_s
+              corrector.replace(expression, correct_style(violation))
+
+              opposite_style_detected
+            end
+          end
+        end
+
+        private
+
+        def preferred_style?(class_reference)
+          class_reference_type = class_reference.type
+
+          return true unless suspicious?(class_reference_type)
+
+          as_style(class_reference_type).equal?(style)
+        end
```

Comment:
Cosmetic. Can also be written as:
```ruby
def opposed_style?(class_reference)
  class_reference_style = REFERENCE_TYPE_STYLES[class_reference.type] # nil, :string, or :constant

  return false unless class_reference_style # Reference is not in the supported style list. No enforcement

  class_reference_style != style
end
```

My nitpick is that `preferred_style?` returns `true` when "reference is not in the supported style list", when it's not semantially a preferred style, it's an ignored style.

This would allow getting rid of the two methods below, and only requires a one-line change in the consumer:
```diff
- return correct_style_detected if preferred_style?(class_reference)
+ return correct_style_detected unless opposed_style?(class_reference)
```

Still, leaving it up to you, the current approach is quite comprehendible, too.

Reply:
Great suggestion. I appreciate that it clarifies the behavior and drops the LOC.

Updated